vllm/core/scheduler.py

```diff
@@ -290,7 +290,8 @@ def __init__(
         # Sequence groups in the SWAPPED state.
         # Contain decode requests that are swapped out.
         self.swapped: Deque[SequenceGroup] = deque()
-
+        # Sequence groups finished since last step iter.
+        self.finished_requests_ids: List[str] = list()
```

Comment:
Could you expand on the comment a bit more, and define the contract between the scheduler and the model for the `finished_request_ids`? I think that we should be able to say that this lets the model know that any state associated with these requests can and must be released after the current step.

Actually I have a question about this: Is it the case that the model holds onto stale state? E.g. if the model generates the last token for a request at inference step N, does it not free that request until inference step N + 1?

Reply:
Done,
RE Question: That's right, we've done that to avoid sending another signal to workers upon finishing requests. And also have as few changes as possible in the workers/engine code, so we've settled on this solution.